streams/jvm/src/test/scala/scalaz/zio/stream/StreamSpec.scala

```diff
@@ -529,4 +535,36 @@ class StreamSpec(implicit ee: org.specs2.concurrent.ExecutionEnv)
         }
     ) must_== Failure(Cause.Fail(e))
   }
+
+  private def bufferStream = prop { list: List[Int] =>
+    unsafeRunSync(
+      Stream
+        .fromIterable(list)
+        .buffer(2)
+        .run(Sink.collect[Int])
+    ) must_== (Success(list))
+  }
+
+  private def bufferStreamError = {
+    val e = new RuntimeException("boom")
+    unsafeRunSync(
+      (Stream.range(0, 10) ++ Stream.fail(e))
+        .buffer(2)
+        .run(Sink.collect[Int])
+    ) must_== Failure(Cause.Fail(e))
+  }
+
+  private def fastProducerSlowConsumer =
+    unsafeRun(
+      for {
+        promise <- Promise.make[Nothing, Ref[List[Int]]]
+        ref     <- Ref.make(List[Int]())
+        _       <- Stream.range(1, 4).mapM(i => ref.update(i :: _)).buffer(2).mapM(_ => IO.never).run(Sink.drain).fork
+        _       <- promise.succeed(ref)
```

Comment:
Almost there. There's no reason to transfer the ref through the promise. You just need to resolve the promise when the element you're interested in passes through the stream. So something like this:
```
_ <- Promise.make[Nothing, Unit]
_ <- Ref.make(List[Int]())
_ <- Stream.range(1, 4).mapM(i => ref.update(i :: _) <* promise.succeed(())).buffer(2).mapM(_ => IO.never).run(Sink.drain).fork
_ <- promise.await
list <- ref.get
```

Reply:
Got it @iravid .. Fixed the review comment.